File: front-end/src/domain/value-object/link.ts
import { LinkName } from './link-name.ts';
import { NodeName } from './node-name.ts';

export class Link {
    private static readonly ERROR_INITIAL_NODE_NOT_DEFINED = "Initial node not defined";
    private static readonly ERROR_FINAL_NODE_NOT_DEFINED = "Final node not defined";
    private static readonly ERROR_LINK_NAME_NOT_DEFINED = "Name not defined";

    private readonly name: LinkName;
    private readonly initialNode: NodeName;
    private readonly finalNode: NodeName;
    private readonly conditionLink: boolean;

    constructor(name: LinkName, initialNode: NodeName, finalNode: NodeName, conditionLink: boolean) {
        this.validateNode(initialNode);
        this.validateFinalNode(finalNode);
        this.validateName(name);
        this.initialNode = initialNode;
        this.name = name;
        this.finalNode = finalNode;
        this.conditionLink = conditionLink;
    }

    private validateName(name: LinkName): void {
        if (!name) {
            throw new Error(Link.ERROR_LINK_NAME_NOT_DEFINED);
        }
    }

    private validateNode(node: NodeName): void {
        if (!node) {
            throw new Error(Link.ERROR_INITIAL_NODE_NOT_DEFINED);
        }
    }

    private validateFinalNode(finalNode: NodeName): void {
        if (!finalNode) {
            throw new Error(Link.ERROR_FINAL_NODE_NOT_DEFINED);
        }
    }

    getName(): LinkName {
        return this.name;
    }

    getInitialNode(): NodeName {
        return this.initialNode;
    }

    getFinalNode(): NodeName {
        return this.finalNode;
    }

    getCondition(): boolean {
        return this.conditionLink;
    }

    updateFinalNode(finalNode: NodeName): Link {
        this.validateFinalNode(finalNode);
        return new Link(this.name, this.initialNode, finalNode, this.conditionLink);
    }

    updateCondition(conditionLink: boolean): Link {
        return new Link(this.name, this.initialNode, this.finalNode, conditionLink);
    }

    updateName(name: LinkName): Link {
        this.validateName(name);
        return new Link(name, this.initialNode, this.finalNode, this.conditionLink);
    }

    updateInitialNode(initialNode: NodeName): Link {
        this.validateNode(initialNode);
        return new Link(this.name, initialNode, this.finalNode, this.conditionLink);
    }


    toString(): string {
        return `Link={name=${this.name}: initialNode=${this.initialNode}, finalNode=${this.finalNode}}`;
    }
}

File: back-end/src/main/java/es/ull/project/domain/valueobject/Link.java
package es.ull.project.domain.valueobject;

/**
 * @brief Class that allows to link a node with another node.
 */
public final class Link {

    private static final String ERROR_INITIAL_NODE_NOT_DEFINED = "Initial node not defined";
    private static final String ERROR_FINAL_NODE_NOT_DEFINED = "Final node not defined";
    private static final String ERROR_LINK_NAME_NOT_DEFINED = "Name not defined";
    private final LinkName name;
    private final NodeName initialNode;
    private final NodeName finalNode;
    private final Boolean conditionLink;

    public Link(LinkName name, NodeName initialNode, NodeName finalNode, Boolean conditionLink) {
        validateName(name);
        validateInitialNode(initialNode);
        validateFinalNode(finalNode);
        this.name = name;
        this.initialNode = initialNode;
        this.finalNode = finalNode;
        this.conditionLink = conditionLink;
    }

    private void validateName(LinkName name) {
        if (name == null) {
            throw new IllegalArgumentException(ERROR_LINK_NAME_NOT_DEFINED);
        }
    }

    private void validateInitialNode(NodeName node) {
        if (node == null) {
            throw new IllegalArgumentException(ERROR_INITIAL_NODE_NOT_DEFINED);
        }
    }

    private void validateFinalNode(NodeName node) {
        if (node == null) {
            throw new IllegalArgumentException(ERROR_FINAL_NODE_NOT_DEFINED);
        }
    }

    public LinkName getName() {
        return this.name;
    }

    public NodeName getInitialNode() {
        return this.initialNode;
    }

    public NodeName getFinalNode() {
        return this.finalNode;
    }

    public Boolean getCondition() {
        return this.conditionLink;
    }

    public Link setName(LinkName name) {
        return new Link(name, this.initialNode, this.finalNode, this.conditionLink);
    }

    public Link setInitialNode(NodeName initialNode) {
        return new Link(name, initialNode, this.finalNode, this.conditionLink);
    }

    public Link setFinalNode(NodeName finalNode) {
        return new Link(name, this.initialNode, finalNode, conditionLink);
    }

    public Link setCondition(Boolean conditionLink) {
        return new Link(name, this.initialNode, this.finalNode, conditionLink);
    }

    @Override
    public boolean equals(Object otherObject) {
        if (this == otherObject) {
            return true;
        }
        if (!(otherObject instanceof Link)) {
            return false;
        }
        Link link = (Link) otherObject;
        return name.equals(link.name) &&
                initialNode.equals(link.initialNode) &&
                finalNode.equals(link.finalNode) &&
                conditionLink.equals(link.conditionLink);
    }

    @Override
    public int hashCode() {
        int result = name.hashCode();
        result = 31 * result + initialNode.hashCode();
        result = 31 * result + finalNode.hashCode();
        result = 31 * result + conditionLink.hashCode();
        return result;
    }

    @Override
    public String toString() {
        return String.format(
                "Link={initialNode=%s, finalNode=%s, name=%s}",
                this.initialNode.getName(), this.finalNode.getName(), name);
    }
}
